### PR TITLE
[Docs][NNCF] TorchFX backend docs

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
https://openvino-doc.iotg.sclab.intel.com/nncf-docs/openvino-workflow/model-optimization-guide/quantizing-models-post-training/basic-quantization-flow.html
### Details:
 - *item1*
 - *...*

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+#! [quantization]
+import torchvision
+import openvino.torch
+model = torchvision.models.resnet50(pretrained=True)
+
+input_fp32 = ... # FP32 model input
+
+with nncf.torch.disable_patching():
+    exported_model = torch.export.export(model, args=(input_fp32,)).module()
```
Comment (alexsu52) [inline docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:25]: I would suggest to use `capture_pre_autograd_graph` and add link to the documentation https://pytorch.org/tutorials/prototype/pt2e_quant_ptq.html to be aligned with PyTorch.

Comment (daniil-lyakhov) [reply docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py]: Done

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+#! [inference]
+import openvino as ov
+
+input_fp32 = ... # FP32 model input
+
+# compile quantized model using torch.compile API
+with nncf.torch.disable_patching():
+    compiled_model_int8 = torch.compile(quantized_model, backend="openvino")
+    # First call compiles an OpenVino model underneath, so it could take longer
```
Comment (rkazants) [inline docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:40]:
```suggestion
    # First call compiles an OpenVINO model underneath, so it could take longer
```

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+...
+
+
+# convert exported Torch model to OpenVINO model
+with nncf.torch.disable_patching():
+    exported_quantized_model = torch.export.export(quantized_model, args=(input_fp32,))
+    ov_quantized_model = ov.convert_model(quantized_model, example_input=input_fp32)
+
+# compile the model to transform quantized operations to int8
+model_int8 = ov.compile_model(ov_quantized_model)
```
Comment (rkazants) [inline docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:52]:
```suggestion
model_int8 = ov.compile_model(ov_quantized_model, 'CPU')
```

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+res = model_int8(input_fp32)
+
```
Comment (rkazants) [inline docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:55]: I think we should have `print` after each variant of inference using `torch.compile` or `OpenVINO`

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    compiled_model_int8 = torch.compile(quantized_model, backend="openvino")
+    # First call compiles an OpenVino model underneath, so it could take longer
+    # than original model call.
+    res = compiled_model_int8(input_fp32)
+...
+
+
+# convert exported Torch model to OpenVINO model
+with nncf.torch.disable_patching():
+    exported_quantized_model = torch.export.export(quantized_model, args=(input_fp32,))
```
Comment (rkazants) [inline docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:48]: I am lost we use both ` torch._export.capture_pre_autograd_graph` and ` torch.export.export`. Do we have any explanation or can we use just one of them for all case

Comment (rkazants) [reply docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:48]: btw, it looks deprecated:
{BA18D293-3D51-41CE-82AA-DE671604F9D4}

Comment (daniil-lyakhov) [reply docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:49]: We are using the same method as in the  [PyTorch 2 Export Post Training Quantization](https://pytorch.org/tutorials/prototype/pt2e_quant_ptq.html). `capture_pre_autograd_graph` to get the `torch.fx.GraphModule` model and `torch.export.export` to export model to `torch.export.ExportedProgram` for ov conversion.

We have two different frontends (`TorchFXPythonDecoder` for `torch.fx.GraphModule`  and `TorchScriptPythonDecoder` for `torch.export.ExportedProgram`). `TorchFXPythonDecoder` can be used only in `torch.compile(...,backend="openvino")` for now, so we need to export `torch.fx.GraphModule` to a `torch.export.ExportedProgram` to make an IR

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py (hunk 3 above)
Comment (daniil-lyakhov) [reply docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:53]: Should we do the same in https://github.com/openvinotoolkit/openvino/blob/master/docs/optimization_guide/nncf/ptq/code/ptq_torch.py, https://github.com/openvinotoolkit/openvino/blob/master/docs/optimization_guide/nncf/ptq/code/ptq_tensorflow.py, https://github.com/openvinotoolkit/openvino/blob/master/docs/optimization_guide/nncf/ptq/code/ptq_openvino.py, https://github.com/openvinotoolkit/openvino/blob/master/docs/optimization_guide/nncf/ptq/code/ptq_onnx.py, https://github.com/openvinotoolkit/openvino/blob/master/docs/optimization_guide/nncf/ptq/code/ptq_aa_openvino.py, https://github.com/openvinotoolkit/openvino/blob/master/docs/optimization_guide/nncf/ptq/code/ptq_aa_onnx.py as well?

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py (hunk 4 above)
Comment (daniil-lyakhov) [reply docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:58]: Done, please check

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py (hunk 5 above)
Comment (daniil-lyakhov) [reply docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:46]: `capture_pre_autograd_graph` is replaced with `export_for_training`

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    # than original model call.
+    res = compiled_model_int8(input_fp32)
+    print(res) # torch.tensor(...)
+...
+
+
+# convert exported Torch model to OpenVINO model
+with nncf.torch.disable_patching():
+    exported_quantized_model = torch.export.export(quantized_model, args=(input_fp32,))
+    ov_quantized_model = ov.convert_model(quantized_model, example_input=input_fp32)
```
Comment (rkazants) [inline docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:47]:
```suggestion
    ov_quantized_model = ov.convert_model(exported_quantized_model, example_input=input_fp32)
```

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+#! [quantization]
+import torchvision
+import openvino.torch
+model = torchvision.models.resnet50(pretrained=True)
+
+input_fp32 = ... # FP32 model input
+
+with nncf.torch.disable_patching():
+    exported_model = torch.export.export_for_training(model, args=(input_fp32,)).module()
+    quantized_model = nncf.quantize(exported_model, calibration_dataset)
```
Comment (rkazants) [inline docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:26]: what is a format of `quantized_model`? torch.nn.Module?

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    # First call compiles an OpenVINO model underneath, so it could take longer
+    # than original model call.
+    res = compiled_model_int8(input_fp32)
+    print(res) # torch.tensor(...)
+...
+
+
+# convert exported Torch model to OpenVINO model
+with nncf.torch.disable_patching():
+    exported_quantized_model = torch.export.export(quantized_model, args=(input_fp32,))
```
Comment (rkazants) [inline docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:46]: here what is a format of `quantized_model`? torch.nn.Module?

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py (hunk 7 above)
Comment (daniil-lyakhov) [reply docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:26]: The format is the [ExportedProgram](https://pytorch.org/docs/stable/export.html#torch.export.ExportedProgram)

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py (hunk 8 above)
Comment (daniil-lyakhov) [reply docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:46]: The format is the [ExportedProgram](https://pytorch.org/docs/stable/export.html#torch.export.ExportedProgram)

Comment (rkazants) [reply docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:46]: according to documentation https://pytorch.org/docs/stable/export.html#torch.export.export, `torch.export.export` anticipates `Module` format

**Review by rkazants — APPROVED** [review]

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+calibration_dataset = nncf.Dataset(calibration_loader, transform_fn)
+#! [dataset]
+
+#! [quantization]
+import torchvision
+import openvino.torch
+model = torchvision.models.resnet50(pretrained=True)
+
+input_fp32 = ... # FP32 model input
```
Comment (alexsu52) [inline docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:22]: I suggest creating a sample input data for exporting the model.

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+def transform_fn(data_item):
+    images, _ = data_item
+    return images
+
+calibration_dataset = nncf.Dataset(calibration_loader, transform_fn)
+#! [dataset]
+
+#! [quantization]
+import torchvision
+import openvino.torch
```
Comment (alexsu52) [inline docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:19]: Why did you add `import openvino.torch`?

Comment (daniil-lyakhov) [reply docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:19]: https://docs.openvino.ai/2024/openvino-workflow/torch-compile.html#how-to-use
image

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py (hunk 9 above)
Comment (daniil-lyakhov) [reply docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:22]: Should I update this line for every example? 
image
image

File: docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+# compile quantized model using torch.compile API
+with disable_patching():
+    compiled_model_int8 = torch.compile(quantized_model, backend="openvino")
+    # OpenVINO backend compiles the model during the first call,
+    # so the first call is expected to be slower than the following calls
+    res = compiled_model_int8(input_fp32)
+
+    # save the model
+    exported_program = torch.export.export(compiled_model_int8, args=(input_fp32,))
```
Comment (alexsu52) [inline docs/optimization_guide/nncf/ptq/code/ptq_torch_fx.py:42]:
```suggestion
    exported_program = torch.export.export(quantized_model, args=(input_fp32,))
```

File: docs/articles_en/openvino-workflow/model-optimization-guide/quantizing-models-post-training/basic-quantization-flow.rst [hunk 12]
```diff
@@ -136,6 +149,17 @@ If you have not already installed OpenVINO developer tools, install it with ``pi
          :language: python
          :fragment:  [inference]
 
+TorchFX models can be compiled and run with OpenVINO optimizations by `torch.compile(..., backend="openvino") <https://docs.openvino.ai/2024/openvino-workflow/torch-compile.html>`__ functionality:
```
Comment (alexsu52) [inline docs/articles_en/openvino-workflow/model-optimization-guide/quantizing-models-post-training/basic-quantization-flow.rst:152]:
```suggestion
TorchFX models can utilize OpenVINO optimizations using `torch.compile(..., backend="openvino") <https://docs.openvino.ai/2024/openvino-workflow/torch-compile.html>`__ functionality:
```

**Review by AlexKoff88 — APPROVED** [review]

**Review by alexsu52 — APPROVED** [review]
